src/_common/generators.ts

```diff
@@ -0,0 +1,92 @@
+/* eslint-disable */
+interface FilepathsOptions {
+  /** Show suggestions with any of these extensions. Do not include the leading dot. */
+  extensions?: string[];
+  /** Show suggestions that include this string */
+  includes?: string;
+  /** Show suggestions where the name exactly matches one of these strings */
+  equals?: string | string[];
+  /** Show suggestions where the name matches this expression */
+  matches?: RegExp;
+  /**
+   * Defines how folders are suggested.
+   *
+   * - **Default:** `always` will always suggest folders.
+   * - `filter` will treat folders like files, filtering based on the name.
+   * - `never` will never suggest folders.
+   */
+  suggestFolders?: "filter" | "always" | "never";
+  /**
+   * Set properties of suggestions of type "file".
+   */
+  editFileSuggestions?: Omit<Fig.Suggestion, "name" | "type">;
+  /**
+   * Set properties of suggestions of type "folder".
+   */
+  editFolderSuggestions?: Omit<Fig.Suggestion, "name" | "type">;
+}
+
+/**
+ * Sugar over using the `filepaths` template with `filterTemplateSuggestions`. If any of the
+ * conditions match, the suggestion will be accepted.
+ *
+ * Basic filepath filters can be replaced with this generator.
+ *
+ * @example
+ * ```
+ * // inside a `Fig.Arg`...
+ * generators: filepaths({ extensions: ["mjs", "js", "json"] });
+ * ```
+ */
+export function filepaths(options: FilepathsOptions): Fig.Generator {
+  const {
+    extensions = [],
+    includes,
+    equals = [],
+    matches,
+    suggestFolders = "always",
+    editFileSuggestions,
+    editFolderSuggestions,
+  } = options;
+  const extensionsSet = new Set(extensions);
+  const equalsSet = new Set(equals);
+  return {
+    template: "filepaths",
+    filterTemplateSuggestions: (suggestions) => {
+      const filtered = suggestions.filter(({ name, type }) => {
+        if (suggestFolders !== "filter" && type === "folder") {
+          return suggestFolders === "always";
+        }
+        if (equalsSet.has(name)) return true;
+        if (matches && matches.test(name)) return true;
+        if (includes && name.includes(includes)) return true;
+        // handle extensions
+        const [_, ...extensions] = name.split(".");
+        if (extensions.length >= 1) {
+          let i = extensions.length - 1;
+          let stackedExtensions = extensions[i];
+          do {
+            if (extensionsSet.has(stackedExtensions)) {
+              return true;
+            }
+            i--;
+            // `i` may become -1 which is not a valid index, but the extensionSet check at the beginning is not run in that case,
+            // so the wrong extension is not evaluated
+            stackedExtensions = [extensions[i], stackedExtensions].join(".");
+          } while (i >= 0);
+        }
```

Comment:
I tried this out, it did end up accessing `extensions[-1]`, but`[undefined, "example"].join(".")` returns `".example"` so it mostly works.

Here's a shorter version, it's a little simpler 🙂

```suggestion
        const lastDot = name.lastIndexOf(".");

        // This loop will never run if there are no extensions
        for (let i = lastDot; i > -1; i--) {
          if (name[i] !== ".") continue;

          // Slice from after the dot to exclude it from the extension
          if (extensionsSet.has(name.slice(i + 1))) return true;
        }
```

---

I tried doing this by looping `lastIndexOf` but it turns out...

```javascript
// Index of the final dot, if we start searching from -1?
".a.b".lastIndexOf(".", -1) === 0
// Of course it's zero... why not, I guess 🥴
```

Reply:
We really don't care that it returns that string because when it is -1 the next check is not performed (the dowhile stops looping). I know it is not intuitive but thats the simplest possible way that does not involve splice.